lib/internal/async_hooks.js

```diff
@@ -243,13 +244,68 @@ function restoreActiveHooks() {
   active_hooks.tmp_fields = null;
 }
 
+const triggerAsyncIdSymbol = Symbol('triggerAsyncId');
+const destroyedSymbol = Symbol('destroyed');
+
+function wrapPromise(promise, parent, silent) {
+  const asyncId = getOrSetAsyncId(promise);
+
+  if (parent) {
+    promise[triggerAsyncIdSymbol] = getOrSetAsyncId(parent);
+    promise.isChainedPromise = true;
+  } else {
+    promise[triggerAsyncIdSymbol] = getDefaultTriggerAsyncId();
+    promise.isChainedPromise = false;
+  }
+
+  if (!silent && initHooksExist()) {
+    emitInitScript(asyncId, 'PROMISE', promise[triggerAsyncIdSymbol], promise);
+  }
+
+  if (destroyHooksExist()) {
+    const destroyed = { destroyed: false };
+    promise[destroyedSymbol] = destroyed;
+    registerDestroyHook(promise, asyncId, destroyed);
+  }
+}
+
+function promiseHook(type, promise, parent) {
+  if (type === 'init' || !promise[async_id_symbol]) {
+    const silent = type !== 'init';
+    if (parent instanceof Promise) {
+      if (!parent[async_id_symbol]) {
+        wrapPromise(parent, null, true);
+      }
+
+      wrapPromise(promise, parent, silent);
+    } else {
+      wrapPromise(promise, null, silent);
+    }
+  }
+
+  if (!promise[async_id_symbol])
+    return;
+
+  switch (type) {
+    case 'before':
+      const triggerId = promise[triggerAsyncIdSymbol];
+      emitBeforeScript(getOrSetAsyncId(promise), triggerId, promise);
+      break;
+    case 'after':
+      emitAfterScript(getOrSetAsyncId(promise));
```

Comment:
Looks like we need to introduce a replacement for this check in `emitAfterScript` (or create a Promise-specific analog of this function):
https://github.com/nodejs/node/blob/468d05aa8544fdaeb11ffcc72263927d2f889ab1/src/async_wrap.cc#L271

Otherwise, if `async_hooks` gets enabled right before the `before` emit, the `executionAsyncResource` stack might get corrupted.